<a href="https://colab.research.google.com/github/shamreen20/Learn-Basic-ML/blob/master/item_Purchase_prediction_using_SVM(Grid_SearchCV)_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Check Item Purchased predictions using SVM without hyper parameter and with hyper parameter(GridDearchCv) **

#Importing libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


**import file**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
df = pd.read_csv("/content/drive/My Drive/social_network_ads.csv")

In [5]:
df.head()

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

**See the dataset is clean or not**

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User ID          400 non-null    int64 
 1   Gender           400 non-null    object
 2   Age              400 non-null    int64 
 3   EstimatedSalary  400 non-null    int64 
 4   Purchased        400 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 15.8+ KB


In [7]:
df.describe()

User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000

**spliting the dataset**

In [11]:
X = df.iloc[:,2:4]

In [12]:
X

Age  EstimatedSalary
0     19            19000
1     35            20000
2     26            43000
3     27            57000
4     19            76000
..   ...              ...
395   46            41000
396   51            23000
397   50            20000
398   36            33000
399   49            36000

[400 rows x 2 columns]

In [13]:
y = df.iloc[:,4]
y

0      0
1      0
2      0
3      0
4      0
      ..
395    1
396    1
397    1
398    0
399    1
Name: Purchased, Length: 400, dtype: int64

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25, random_state=50)

In [16]:
from sklearn.preprocessing import StandardScaler

In [17]:
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

**Simply train the model by SVM(search vector machine) without hyper parameter** 

In [18]:
from sklearn.svm import SVC

In [20]:
classifier = SVC(kernel="linear", random_state=0)
classifier.fit(X_train,y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=0, shrinking=True, tol=0.001,
    verbose=False)

**predict the result**

In [21]:
y_pred = classifier.predict(X_test)

In [22]:
from sklearn.metrics import confusion_matrix, classification_report

In [23]:
confusion_matrix(y_test,y_pred)

array([[68,  2],
       [13, 17]])

In [25]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.84      0.97      0.90        70
           1       0.89      0.57      0.69        30

    accuracy                           0.85       100
   macro avg       0.87      0.77      0.80       100
weighted avg       0.86      0.85      0.84       100



***here we see without hyper parameter we got 0.85 accuray***


**Now we gonna use Grid searchCV to find the best model and the best parameters**

In [26]:
from sklearn.model_selection import GridSearchCV

In [28]:
parameters = [{'C':[1,10,100,1000],'kernel':['linear']},
              {'C':[1,10,100,1000], 'kernel':['rbf'], 'gamma':[0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9]}]

In [29]:
grid_search = GridSearchCV(estimator = classifier,
                          param_grid = parameters,
                           scoring='accuracy',
                           cv=10,n_jobs=-1)

In [30]:
grid_search = grid_search.fit(X_train,y_train)

In [31]:
grid_search.best_params_

{'C': 100, 'gamma': 0.8, 'kernel': 'rbf'}

In [32]:
classifier = SVC(kernel="rbf",gamma=0.7)

In [33]:
classifier.fit(X_train,y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.7, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

**Predicting the result**

In [34]:
from sklearn.metrics import confusion_matrix, classification_report

In [35]:
y_pred = classifier.predict(X_test)

In [36]:
confusion_matrix(y_test,y_pred)

array([[67,  3],
       [ 3, 27]])

In [37]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.96      0.96      0.96        70
           1       0.90      0.90      0.90        30

    accuracy                           0.94       100
   macro avg       0.93      0.93      0.93       100
weighted avg       0.94      0.94      0.94       100



***Here, by using GridSearchCV the accuracy score increases to 0.94 ***